### PR TITLE
Install serverless from midstream

### PR DESCRIPTION
**Pull request by aliok**
JIRA: https://issues.redhat.com/browse/MGDOBR-1127

Done:
- Install serverless from midstream
- Install serverless operator in another namespace (`openshift-serverless`) than `openshift-operators`. This is the recommended way. Also, existing installations are migrated to use this namespace per the [migration document](https://github.com/openshift-knative/serverless-operator/blob/main/docs/migrating-from-product-to-midstream.md).
- Make some shell functions accept an optional namespace argument

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [ ] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [ ] The layers in the `kustomize` folder have been updated accordingly.
- [ ] All new functionality is tested
- [ ] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

File: scripts/dependencies/common_tools.sh [hunk 1]
```diff
@@ -34,36 +34,37 @@ function waitForSuccess() {
     done
 }
 
-# Installs an operator according to given subscription resource and waits, params: operatorSubscriptionResource
+# Installs an operator according to given subscription resource and waits, params: operatorSubscriptionResource, namespace (defaults to openshift-operators)
```
Comment (aliok) [inline scripts/dependencies/common_tools.sh:37]: Added this argument as serverless operator is no longer installed in `openshift-operators` namespace anymore.

File: scripts/dependencies/common_tools.sh [hunk 2]
```diff
@@ -34,36 +34,37 @@ function waitForSuccess() {
...
 # Example:
 #  install_operator_and_wait serverless/serverlessSub.yaml
 #
 function install_operator_and_wait()
 {
   local resource=$1
+  local namespace=${2:-openshift-operators}
```
Comment (aliok) [inline scripts/dependencies/common_tools.sh:44]: I don't like magic defaulting in general, and I don't like it even more in shell scripts. I personally prefer not do any defaulting and gracefully failing if argument is missing, but that means changing everywhere where this function is used.

Let me know what your preference is and I can change.

File: scripts/dependencies/install_dependencies.sh [hunk 3]
```diff
@@ -3,8 +3,13 @@
...
 
 echo "Installing Serverless Operator"
-# based on https://docs.openshift.com/container-platform/4.11/serverless/install/install-serverless-operator.html
-install_operator_and_wait serverless/serverlessSub.yaml
+# based on https://github.com/openshift-knative/serverless-operator/blob/main/docs/install-midstream.md
+waitForSuccess 10 oc apply -f serverless/serverless-system-namespace.yaml
+waitForSuccess 10 oc apply -f serverless/serverless-operator-group.yaml
+waitForSuccess 10 oc apply -f serverless/serverless-catalog-source.yaml
+# TODO: do we want to wait until catalog source is ready?
+# oc wait catalogsources -n openshift-marketplace serverless-operator-v1-24-0 --for=jsonpath='{.status.connectionState.lastObservedState}'="READY" --timeout=5m
```
Comment (aliok) [inline scripts/dependencies/install_dependencies.sh:11]: I haven't seen these kind of waits for other resources and I am not 100% sure if we're good w/o it.

**Comment by aliok** [comment]
Pushing for early feedback.

BTW, would you like me to create a JIRA first?

**Review by r00ta — APPROVED** [review]

**Comment by robpblake** [comment]
@aliok This can done as part of https://issues.redhat.com/browse/MGDOBR-1127 👍

**Review by r00ta — COMMENTED** [review]
Hey @aliok , it seems something has failed 
```bash
./install_dependencies.sh 
Installing Serverless Operator
namespace/openshift-serverless created
operatorgroup.operators.coreos.com/serverless created
catalogsource.operators.coreos.com/serverless-operator-v1-24-0 created
Applying serverless/serverlessSub.yaml
subscription.operators.coreos.com/openshift-serverless-operator-subscription created
Waiting for operator installation plan
Trying again. Try #0
Trying again. Try #1
Trying again. Try #2
Trying again. Try #3
Trying again. Try #4
Trying again. Try #5
Trying again. Try #6
Trying again. Try #7
Trying again. Try #8
Trying again. Try #9
Trying again. Try #10
Trying again. Try #11
Trying again. Try #12
Trying again. Try #13
Trying again. Try #14
Trying again. Try #15
Trying again. Try #16
Trying again. Try #17
Trying again. Try #18
Trying again. Try #19
Trying again. Try #20
Trying again. Try #21
Trying again. Try #22
Trying again. Try #23
Trying again. Try #24
Trying again. Try #25
Trying again. Try #26
Trying again. Try #27
Trying again. Try #28
Trying again. Try #29
Failed! waitForOpResult reached max retry count 30.
Operator installation plan found: 
Approving operator installation plan
error: resource(s) were provided, but no name, label selector, or --all flag specified
Trying again. Try #0
error: resource(s) were provided, but no name, label selector, or --all flag specified
Trying again. Try #1
error: resource(s) were provided, but no name, label selector, or --all flag specified
Trying again. Try #2
error: resource(s) were provided, but no name, label selector, or --all flag specified
Failed! Wait for success reached max retry count 3.
```

The error in the cluster is 
```
[failed to populate resolver cache from source redhat-operators/openshift-marketplace: failed to list bundles: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 172.30.248.200:50051: connect: connection refused", failed to populate resolver cache from source certified-operators/openshift-marketplace: failed to list bundles: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 172.30.223.131:50051: connect: connection refused"]
```

**Comment by aliok** [comment]
@r00ta 

This is the output I get on a fresh OCP cluster:
```
❯ ./install_dependencies.sh                                                                                                                                                                                                                                                                                          ─╯
Installing Serverless Operator
namespace/openshift-serverless created
operatorgroup.operators.coreos.com/serverless created
catalogsource.operators.coreos.com/serverless-operator-v1-24-0 created
Applying serverless/serverlessSub.yaml
subscription.operators.coreos.com/openshift-serverless-operator-subscription created
Waiting for operator installation plan
Trying again. Try #0
Trying again. Try #1
Operator installation plan found: install-mrlsz
Approving operator installation plan
installplan.operators.coreos.com/install-mrlsz patched
Waiting for operator CSV
Operator CSV found: serverless-operator.v1.24.0
Error from server (NotFound): clusterserviceversions.operators.coreos.com "serverless-operator.v1.24.0" not found
Trying again. Try #0
Error from server (NotFound): clusterserviceversions.operators.coreos.com "serverless-operator.v1.24.0" not found
Trying again. Try #1
Error from server (NotFound): clusterserviceversions.operators.coreos.com "serverless-operator.v1.24.0" not found
Trying again. Try #2
Pending
InstallReady
Waiting for status.phase=Succeeded
clusterserviceversion.operators.coreos.com/serverless-operator.v1.24.0 condition met
networkpolicy.networking.k8s.io/eventing-webhook created
networkpolicy.networking.k8s.io/kafka-broker-receiver created
knativeeventing.operator.knative.dev/knative-eventing created
knativekafka.operator.serverless.openshift.io/knative-kafka created
```

I am just thinking if there can be any limitations in the OCP cluster you're using. Is there any special config?

**Comment by aliok** [comment]
Ok, tried on a fresh cluster from cluster-bot. Took longer, but succeeded:

```
❯ ./install_dependencies.sh                                                                                                                                                                                                                                                                                          ─╯
Installing Serverless Operator
namespace/openshift-serverless created
operatorgroup.operators.coreos.com/serverless created
catalogsource.operators.coreos.com/serverless-operator-v1-24-0 created
Applying serverless/serverlessSub.yaml
subscription.operators.coreos.com/openshift-serverless-operator-subscription created
Waiting for operator installation plan
Trying again. Try #0
Trying again. Try #1
Trying again. Try #2
Trying again. Try #3
Operator installation plan found: install-cr9g2
Approving operator installation plan
installplan.operators.coreos.com/install-cr9g2 patched
Waiting for operator CSV
Operator CSV found: serverless-operator.v1.24.0
Error from server (NotFound): clusterserviceversions.operators.coreos.com "serverless-operator.v1.24.0" not found
Trying again. Try #0
Error from server (NotFound): clusterserviceversions.operators.coreos.com "serverless-operator.v1.24.0" not found
Trying again. Try #1
Error from server (NotFound): clusterserviceversions.operators.coreos.com "serverless-operator.v1.24.0" not found
Trying again. Try #2
Pending
Pending
Waiting for status.phase=Succeeded
clusterserviceversion.operators.coreos.com/serverless-operator.v1.24.0 condition met
networkpolicy.networking.k8s.io/eventing-webhook created
networkpolicy.networking.k8s.io/kafka-broker-receiver created
knativeeventing.operator.knative.dev/knative-eventing created
knativekafka.operator.serverless.openshift.io/knative-kafka created
```

Maybe it was a temporary issue?

Overall, this line should check if there's actually an installPlan found and fail gracefully: https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/dbde5a96443d5c21d5728eb6e1953f07cc0eefb6/scripts/dependencies/common_tools.sh#L52

That's not part of this PR though.

**Comment by pierDipi** [comment]
it's weird that it is failing on the marketplace source `source redhat-operators/openshift-marketplace`, I'd expect it to fail on our created source. :thinking:

**Comment by aliok** [comment]
> it's weird that it is failing on the marketplace source source redhat-operators/openshift-marketplace, I'd expect it to fail on our created source. 🤔

Good point. Then, it could've been any operator from redhat-operators source, right? Not sure if it is about serverless here. Looks more like a flake in the catalogSource image

**Comment by r00ta** [comment]
Yes might be. Let's move forward

**Comment by aliok** [comment]
@r00ta This is good to merge

**Comment by r00ta** [comment]
@manstis @robpblake @sutaakar could any of you approve this please

**Review by sutaakar — APPROVED** [review]

**Comment by aliok** [comment]
About this error:

```
[failed to populate resolver cache from source redhat-operators/openshift-marketplace: failed to list bundles: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 172.30.248.200:50051: connect: connection refused", failed to populate resolver cache from source certified-operators/openshift-marketplace: failed to list bundles: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 172.30.223.131:50051: connect: connection refused"]
```

@maschmid pointed me to https://issues.redhat.com/browse/CLOUDDST-15927

> there seems to be an issue with this image in redhat registry https://issues.redhat.com/browse/CLOUDDST-15927

**Pull request merged**